components/annot-vocabulary.vue

```diff
@@ -202,21 +222,53 @@
                 return placeHolderText;
             },
 
+            saveButtonEnabled() {
+
+                for ( const [columnName, columnMappings] of Object.entries(this.vocabularyMapping) ) {
+                    console.log("val", columnName, columnMappings)
+                    for ( const [uniqueValue, mappedValue] of Object.entries(columnMappings) ) {
+                        // The first time we find any mapped value that is neither non-empty nor missing, we return
+                        // a status of false. Otherwise, we return true.
+                        if (
+                            // We use negate the first conditional tuple so that lazy evaluation of the two conditionals
```

Comment:
The comment is not referring to the if statement but to the logic inside the condition. Could you maybe give a reason why you think we should not put comments inside if statements?

Reply:
The short answer to this is precedent - it's just not common practice to do so. The better answer though is that the condition should be rewritten for clarity and in doing so likely not need as much commentary.

I believe this original comment

```
// The first time we find any mapped value that is neither non-empty nor missing, we return
// a status of false. Otherwise, we return true.

if (
     // We use negate the first conditional tuple so that lazy evaluation of the two conditionals
     // joined by && will skip the second conditional if uniqueValue is null. This is important
     // because null doesn't have a .trim() method.
     ! ( mappedValue  !== null && mappedValue.trim().length > 0 ) &&
     ( ! this.isMissingValue(columnName, uniqueValue) )
) {
    return false;
}
```

Can also be written this way:

```
// If there is any mapped value that is empty or unique value that is not missing, we return false
if ( (mappedValue !== null && mappedValue.trim().length === 0) || 
      !this.isMissingValue(columnName, uniqueValue) ) {

    return false;
}
```

The reasoning for this as follows. If we map out how your original condition could succeed, we get the following:

```
mappedValue !== null && mappedValue.trim().length === 0 (T and F)
OR
mappedValue === null --> short-circuit out of evaluating second condition (F)

AND

this.isMissingValue(columnName, uniqueValue) === false
```

@surchs Can you provide some feedback here as to if I have correctly understood the intent of this condition?

Reply:
Ah, hold on. Now I confused myself (goes to show that the if clause needs cleaning). The idea is that we want to turn on/make available the NEXT button when all unique values have been either mapped or declared missing.


So for a `uniqueValue`, if
the corresponding `mappedValue` is NOT set (i.e. is `null`) OR is empty (i.e. is an empty string with length zero)
AND
the `uniqueValue` has not been marked as a missingValue
THEN
this tells me that at least one value remains unmapped and the button shouldn't be enabled.

(this is the intent, not the literal code)

The problem is that we can't write it like that because if
`mappedValue === null` is true, then we'll go on to evaluate `mappedValue.trim(appedValue !== null).length > 0` and then crash because a null object doesn't have a `trim()` method.

Your version does a different thing:
```javascript
if ( (mappedValue !== null && mappedValue.trim().length === 0) || 
      !this.isMissingValue(columnName, uniqueValue) ) {
```
for a `uniqueValue`, if
the corresponding `mappedValue` IS set (i.e. is NOT `null`) BUT is empty (i.e. is an empty string with length zero)
OR
the `uniqueValue` has not been marked as a missingValue
THEN
we consider the annotation not complete and the button to not be enabled. But this is not correct, the `OR` here is wrong.

Reply:
Haha...I've had to think and rethink about this too.

Since you say, that:

>The idea is that we want to turn on/make available the NEXT button when all unique values have been either mapped or declared missing.

How about this?

````
if ( mappedValue === null ||
      (mappedValue !== null && mappedValue.trim().length > 0) ||
     !this.isMissingValue(columnName, uniqueValue ) {
```

So if a) mappedValue is null or b) mapped value is empty or c) uniqueValue is not has been marked as not missing.
I think that accords with what you were saying.

Also FYI, though maybe not useful in this case, JavaScript now has the optional chaining `?.` operator to overcome null/undefined fields on an object being accessed and handled in a smoother fashion: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Reply:
This chaining operator looks exactly like what we could use here. I was looking for that. It's like the python dict.get() method in that sense, yes? 

Regarding the if statement, I do like your new version better because it is clearer. But it is still not exactly what we want.
A value is NOT ANNOTATED when it is:

- not yet set (still at default value of `null`) or emtpy (only whitespace string)  AND ALSO
- the value has not been declared missing.

The AND is because we want to enable completion of the annotation when some values are declared missing (and therefor not annotated). So the "go" condition for annotation done is: everything is either annotated (not null and not empty) or declared missing. 

Since we both are struggling to keep this sorted in our heads, I am wondering if we shouldn't stop trying to be smart here and just write the logic exactly like that. I think in that case we'd go back to the functional form and do something like:

```javascript
return Object.entries(this.vocabularyMapping).every([columnName, columnMappings] => {
return columnMappings.every([uniqueValue, mappedValue] => {
return (
! ( mappedValue !== null && mappedValue.trim().length > 0) ||
this.isMissingValue(columnName, uniqueValue)
)
}
)
}
)
```

Reply:
So now we would have to actually go and look at every value in every column to return true. But the logic is actually what we mean, rather than some fancy early exit stuff that has us scratching our heads again or writing a half-page comment to explain why we did it this way

Reply:
One more go at it before returning to the functional style.

Is this it?

```
if ( ( mappedValue === null || (mappedValue !== null && mappedValue.trim().length > 0) ) &&
     !this.isMissingValue(columnName, uniqueValue ) {

    return false;
}
```

Reply:
Aha, yes, I think this is going to be it, with one minor change (`length > 0` -> `length === 0`):

```javascript
if ( ( mappedValue === null || (mappedValue !== null && mappedValue.trim().length === 0) ) &&
     !this.isMissingValue(columnName, uniqueValue ) {

    return false;
}

```
And then maybe add a comment on why we cannot just do
````javascript
if ( ( mappedValue === null || mappedValue.trim().length === 0 ) && ...
```
so that we don't accidentally "improve" it later and get surprised when things break.

Very cool, thanks for thinking through this!

Reply:
It had to get done! Good talking through.

Reply:
Actually @surchs 

```
if ( ( mappedValue === null || mappedValue.trim().length === 0 ) && ...
```

should work given that if if `mappedValue` is `null` it won't check the second condition once `mappedValue === null` evaluates to true. But of course, `mappedValue !== null` can be left in for clarity even if it's redundant

Reply:
Ah, that's right. I didn't think about this because now we have turned the statement from `!==` to `===`. OK, in that case I think it is really clear, we might not even need a comment. Look at that!

Reply:
Excellent work. Exactly where we want it to be.